ffi_derive/src/enum_ffi.rs

```diff
@@ -30,11 +34,68 @@ pub(super) fn build(module_name: &Ident, type_name: &Ident, out_dir: &str) -> To
             use super::*;
 
             #[no_mangle]
-            pub unsafe extern "C" fn #fn_name(data: #type_name) {
+            pub unsafe extern "C" fn #free_fn_name(data: #type_name) {
                 let _ = data;
             }
 
             declare_value_type_ffi! { #type_name }
         }
     }
 }
+
+pub(super) fn complex(
+    module_name: &Ident,
+    type_name: &Ident,
+    derive: &DataEnum,
+    alias_modules: &[String],
+    required_imports: &[Path],
+    out_dir: &str,
+) -> TokenStream {
+    let variants = derive
+        .variants
+        .iter()
+        .map(|variant| {
+            let other_variants = derive
+                .variants
+                .iter()
+                .cloned()
+                .filter_map(|other_variant| {
+                    if &other_variant == variant {
+                        None
+                    } else {
+                        Some(other_variant.ident)
+                    }
+                })
+                .collect();
+            let fields = ffi_internals::struct_internals::field_ffi::fields_for_variant(
+                type_name,
+                alias_modules,
+                &variant.ident,
+                &variant.fields,
+                other_variants,
+            );
+            VariantFFI {
+                ident: &variant.ident,
+                fields,
+            }
+        })
+        .collect();
+
+    let enum_ffi = EnumFFI {
+        module_name,
+        type_name,
+        variants,
+        alias_modules,
+        required_imports,
+    };
+    let file_name = format!("{}.swift", type_name.to_string());
+    ffi_internals::write_consumer_file(
+        &file_name,
+        String::from(&ComplexConsumerEnum::new(&enum_ffi) as &dyn ConsumerType),
```

Comment:
I'm getting this warning on the explicit cast
```
warning: trivial cast: `&ComplexConsumerEnum<'_>` as `&dyn ConsumerType`
  --> /Users/nicholas.smillie/repos/ffi_common/ffi_derive/src/enum_ffi.rs:94:22
   |
94 |         String::from(&ComplexConsumerEnum::new(&enum_ffi) as &dyn ConsumerType),
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: cast can be replaced by coercion; this might require a temporary variable
```

Which is reasonable sounding, but the only way I've been able get coercion to do it for me is with
```
let complex_consumer_enum = ComplexConsumerEnum::new(&enum_ffi);
let consumer: &dyn Consumer = complex_consumer_enum;
String::from(consumer)
```

And I'm not sure two extra temporary variables is worth avoiding writing the trivial cast. Is there a better way to tell it to use the trait's implementation, or should I just disable the lint for this one?